### PR TITLE
[JVM] Fix the nop removal optimization.

### PR DESCRIPTION
**Pull request by madsager**
Never remove a nop if that would cause a line number to move
across a local lifetime boundary.

This fixes KT-42725.

### CODE REVIEW

File: compiler/testData/codegen/bytecodeText/coercionToUnitOptimization/nopInlineFuns.kt [hunk 1]
```diff
@@ -23,4 +23,4 @@ fun simpleFunVoid(f: () -> Unit): Unit {
     return f()
 }
 
-// 3 NOP
```
Comment (madsager) [inline compiler/testData/codegen/bytecodeText/coercionToUnitOptimization/nopInlineFuns.kt:26]: Stepping through this code works beautifully without the nops, so I don't think any nops are needed. Do you see any reason there would have to be nops in the generated code for this?

**Comment by madsager** [comment]
I should probably add a bit more of an explanation for this. The issue that there is a line number on a return instruction in an inline function in the standard library `filter` function:

```
L1:
   // Code for the inline function.
L2:
   LINENUMBER 6
   RETURN
L3:
```

The locals table says that the inlining variable `$i$f$filter` goes from label 1 to label 3.

Now, when inlining this code, we remove the `return` and replace it with a `nop`:

```
L1:
   // Code for the inline function.
L2:
   LINENUMBER 6
   NOP
L3:
  CHECKCAST 
  // Code that uses the result.
```

The locals table still says that the inlining variable goes from label 1 to label 3.

Unfortunately, the nop removal algorithm does not work correctly for this case and it ends up removing the `nop`. That puts the line number on the `checkcast` instruction. However, that is now outside of the range of the inlining variable which leads the debugger to break there. Because it is outside of the range of the inlining variable, the debugger does not filter out the break as being from the standard library. So using step-over, we actually end up stepping into the `filter` library inline function.

File: compiler/testData/codegen/bytecodeText/coercionToUnitOptimization/nopInlineFuns.kt (hunk 1 above)
Comment (dnpetrov) [reply compiler/testData/codegen/bytecodeText/coercionToUnitOptimization/nopInlineFuns.kt:26]: I think that's fine. @yanex could you please take a look?

**Comment by dnpetrov** [comment]
Overall, looks good to me. I'll run tests and push manually.

**Comment by dnpetrov** [comment]
Pushed manually.

318d11dcecdac5c2bc137e5a1b3abd5faba7f9ce
0505bd958a835a3be45a6224d734b1f2689efcdd

**Pull request closed without merging**